### ISSUE TITLE
Reproducing StackLLaMA

### ISSUE THREAD

**Issue by mnoukhov** [huggingface/trl#401]
I've reproduced the whole StackLLaMA pipeline using the changes in #398 #399 #400 

Here is the [corresponding wandb report](https://wandb.ai/mnoukhov/trl/reports/StackLLaMA-Repro--Vmlldzo0NTM1MDk2)

A couple notes:
- As my base llama I used `huggyllama/llama-7b` 
- My supervised ft run was better the in the blog post, reaching a lower ppl
- My reward modelling run was worse than the blog post (67%) and only reached 63% after one epoch. So I ran it for two epochs and got ~66% which I felt was sufficient
- The RL training curves look very similar. I found that I could achieve similar performance with a lower KL coefficient (0.02) in less training time, 600 epochs vs 1200 but still have the original KL coefficient run (0.2)

I've also published my adapter weights on the hub 
https://huggingface.co/mnoukhov/llama-7b-se-peft
https://huggingface.co/mnoukhov/llama-7b-se-rm-peft
https://huggingface.co/mnoukhov/llama-7b-se-rl-peft

Use the `merge_peft` script in #398 to merge `huggyllama/llama-7b` and `llama-7b-se-peft` to make `llama-7b-se`
Then merge `llama-7b-se` with `llama-7b-se-rm-peft` to make the reward model and `llama-7b-se-rl-peft` to the make StackLLaMA

**Comment by mnoukhov** [comment]
Sorry, I moved the runs to the workspace so the graphs should be fixed.

My libraries are 
```
accelerate==0.18.0
evaluate==0.4.0
huggingface-hub==0.13.3
torch==2.0.0
transformers==4.28.1
```

and the latest version of trl built from source

**Comment by dh2shin** [comment]
Hi @mnoukhov , could you explain when/what exactly to merge? I'm following the [readme](https://github.com/lvwerra/trl/tree/main/examples/stack_llama/scripts) and would really appreciate your help.

**Comment by mnoukhov** [comment]
That's correct. 

base: `huggyllama` + `llama-7b-se-peft` = `llama-7b-se`
base `llama-7b-se` + `llama-7b-se-rm-peft` = `llama-7b-se-rm` 
base `llama-7b-se` + `llama-7b-se-rl-peft` = `llama-7b-se-rl`

**Comment by dh2shin** [comment]
Do you mind sharing the arguments / shell script you used for each step? I'm using what's listed on the [repo](https://github.com/lvwerra/trl/tree/main/examples/stack_llama/scripts), having memory requirement issues, which seems odd given peft+lora.

**Comment by mnoukhov** [comment]
I use the same hyperparameters as those listed with the slight change that I am running on 4 GPUs instead of 8 so I change gradient accumulation steps from 4 to 8. I find that I need ~40GB of GPU memory for the RL finetuning step but it goes back and forth and can get as high as 60 per GPU. Currently working on #436 which should reduce memory requirements to allow for 32GB GPU training

**Comment by dh2shin** [comment]
I'm experiencing similar effects, about ~40GB of GPU memory is needed for the second step of training the Reward Model too, right?

**Comment by mnoukhov** [comment]
You can check the exact memory and compute things by looking at the runs linked in the wandb report (e.g. [my RLHF run](https://wandb.ai/mnoukhov/trl/runs/212500ae7be6c616547edda08e29c2ef) shows the memory consumption in the "System" charts)

Given that I've essentially repro'd the results and my PRs have been merged, I'm closing this issue and continuing in #471 with repro using the more compute-efficient multi-adapter paradigm. Feel free to keep commmenting about the repro and I'll try to respond.

**Comment by mnoukhov** [comment]
None of the messages are related. There are other issues related to instability in training, and without more info it's hard to diagnose the problem. If you want some advice specific to your situation it would be helpful to know lots of things like reward, KL, etc... 

If you just want to try some things out #462 found that setting a larger minibatch size and larger target KL could help

**Comment by dh2shin** [comment]
@mnoukhov Hi Michael, when using the reward model to build out the sentiment pipeline in rl_training.py, the output rewards/scores vary drastically depending on the batch_size used in sent_kwargs. I am wondering if you have investigated this issue more in-depth.

**Comment by wangzhao88** [comment]
hi, i use your rl model[base llama-7b-se + llama-7b-se-rl-peft = llama-7b-se-rl] to test the superglue by https://github.com/EleutherAI/lm-evaluation-harness, the result is follow:
| Task    | Version | Metric | Value  |     | Stderr |
| boolq     | 1        | acc    | 0.7642 | ±   | 0.0074 |
| cb          | 1        | acc    | 0.5536  | ±  | 0.0670 |
|               |           | f1       | 0.4248 |      |              |
| copa      | 0       | acc    | 0.8800 | ±   | 0.0500 |
| multirc  | 1        | acc    | 0.0084  | ±   | 0.0018 |
| record   | 0       | f1       | 0.9119   | ±   | 0.0032 |
|               |          | em     | 0.9044  | ±    | 0.0032 |
| rte         | 0       | acc    | 0.6282 | ±    | 0.0301 |
| wic        | 0       | acc    | 0.4953 | ±    | 0.0198 |
| wsc       | 0       | acc    | 0.5673 | ±    | 0.0474 |

however, when i train my own rl model by:
'''
accelerate launch --multi_gpu --num_machines 1  --num_processes 8 rl_training.py --model_name=【your se model】 --reward_model_name=【your rm model】 --adafactor=False --tokenizer_name=【your se model】 --save_freq=100 --output_max_length=128 --batch_size=8 --gradient_accumulation_steps=8 --batched_gen=True --ppo_epochs=4 --seed=0 --learning_rate=1.4e-5 --early_stopping=True --output_dir=llama-se-rl-finetune-128-8-8-1.4e-5_adam
'''

after one day, the result of my own rl model(llama-se-rl-finetune-128-8-8-1.4e-5_adamstep_1300) is:
| Task    | Version | Metric | Value  |     | Stderr |
| boolq     | 1       | acc    | 0.3783 | ±   | 0.0085 |
| cb          | 1        | acc    | 0.4107  | ±  | 0.0663 |
|               |           | f1       | 0.1941  |      |              |
| copa      | 0       | acc    | 0.5500 | ±   | 0.0500 |
| multirc  | 1        | acc    | 0.0031  | ±   | 0.0018 |
| record   | 0       | f1       | 0.1186   | ±   | 0.0032 |
|               |          | em     | 0.1151   | ±    | 0.0032 |
| rte         | 0       | acc    | 0.5271  | ±    | 0.0301 |
| wic        | 0       | acc    | 0.5000 | ±    | 0.0198 |
| wsc       | 0       | acc    | 0.6346 | ±    | 0.0474 |

is means that the training of model is failed?

**Comment by lvwerra** [comment]
Can you share the logs from the RL training? E.g. mean rewards and objective/kl are usually helpful metrics to look at to see if the model learned something.

**Comment by wangzhao88** [comment]
> Can you share the logs from the RL training? E.g. mean rewards and objective/kl are usually helpful metrics to look at to see if the model learned something.

i will share the logs soon~

**Comment by lvwerra** [comment]
Looks like there was an issue at step ~50: the reward went down significantly. Could you try with a different seed or lower learning rate? Also we added some stability measures in the latest release so try updating `trl`.

**Comment by wangzhao88** [comment]
Hello! Here is the latest log: https://wandb.ai/630191510/trl/runs/kj2kkbq9?workspace=user-630191510

The loss curve looks normal, and the accuracy in SuperGLUE is also normal.

I suppose the key difference in ppo_trainer.py between the two branches is crucial.

**Comment by lvwerra** [comment]
That's great! So updating helped?

**Comment by wangzhao88** [comment]
Here is the log:
[Link to the old TRL run with seed 2023](https://wandb.ai/630191510/trl/runs/f6lzk9d2?workspace=user-630191510)
[Link to the new TRL run with seed 0](https://wandb.ai/630191510/trl/runs/ur6maq2k?workspace=user-630191510)

It appears that trying different seeds or updating the code could potentially improve the training of PPO.

**Comment by zhangfudiyi** [comment]
it‘s great jobs
